pkg/provider/provider.go

```diff
@@ -621,3 +636,83 @@ func validateFileName(fileName string) error {
 	}
 	return nil
 }
+
+type node struct {
+	cert     *x509.Certificate
+	parent   *node
+	isParent bool
+}
+
+func fetchCertChains(data []byte) ([]byte, error) {
+	var newCertChain []*x509.Certificate
+	var pemData []byte
+	nodes := make([]*node, 0)
+
+	for {
+		// decode pem to der first
+		block, rest := pem.Decode(data)
+		data = rest
+
+		if block == nil {
+			break
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return pemData, err
+		}
+		nodes = append(nodes, &node{
+			cert:     cert,
+			parent:   nil,
+			isParent: false,
+		})
+	}
+
+	// at the end of this computation, the output will be a single linked list
+	// the tail of the list will be the root node (which has no parents)
+	// the head of the list will be the leaf node (whose parent will be intermediate certs)
+	// (head) leaf -> intermediates -> root (tail)
+	for i := range nodes {
+		for j := range nodes {
+			// ignore same node to prevent generating a cycle
+			if i == j {
+				continue
+			}
+			// if ith node AuthorityKeyId is same as jth node SubjectKeyId, jth node was used
+			// to sign the ith certificate
+			if string(nodes[i].cert.AuthorityKeyId) == string(nodes[j].cert.SubjectKeyId) {
+				nodes[j].isParent = true
+				nodes[i].parent = nodes[j]
+				break
+			}
+		}
+	}
+
+	var leaf *node
+	for i := range nodes {
+		if !nodes[i].isParent {
+			// this is the leaf node as it's not a parent for any other node
+			// TODO (aramase) handle errors if there are more than 1 leaf nodes
+			leaf = nodes[i]
+			break
+		}
+	}
+
+	if leaf == nil {
+		return nil, fmt.Errorf("no leaf found")
+	}
+
+	// iterate through the directed list and append the nodes to new cert chain
+	for leaf != nil {
+		newCertChain = append(newCertChain, leaf.cert)
+		leaf = leaf.parent
+	}
+
+	for _, cert := range newCertChain {
```

Comment:
Do we need to validate that all three parts are included in newCertChain?

Reply:
@ritazh Do you mean to validate that there is a leaf, an intermediate, and a CA.  The code should not do that as not all PEM files will have all three and they are not required to.  They may only contain a leaf certificate.

Reply:
> The code should not do that as not all PEM files will have all three and they are not required to.

Gotcha!